### PR TITLE
add rules model

### PR DESCRIPTION
**Pull request by cwen0**
@zhouqiang-cl @foreverbell PTAL

### CODE REVIEW

File: oasis/models/rules.py [hunk 1]
```diff
@@ -2,15 +2,168 @@
...
+                rules_list.append({"expr": rule, "feature": feature,
+                                   "operator": operator, "value": value})
 
+            self.rules[metric] = rules_list
 
 
+def check_is_triggered(left_value, operator, right_value):
+    is_triggered = False
+    return {
+        "==": lambda is_triggered: abs(left_value-right_value) < 0.0001,
```
Comment (foreverbell) [inline oasis/models/rules.py]: where to use?

Comment (cwen0) [reply oasis/models/rules.py]: This function is used in rules.py/L76

Comment (foreverbell) [reply oasis/models/rules.py]: I mean `is_triggered`?

File: oasis/models/model.py [hunk 2]
```diff
@@ -2,23 +2,66 @@
...
-        logger.info("running calculation model")
+        logger.info("run model")
 
     def close(self):
-        logger.info("close model")
+        with self.lock:
+            self._exit = True
+            self.event.set()
+            for key in self.threads:
+                self.threads[key].join()
```
Comment (zhouqiang-cl) [inline oasis/models/model.py]: Did we need a timeout for join

File: oasis/models/model.py [hunk 3]
```diff
@@ -2,23 +2,66 @@
...
 
 
+class Config(object):
+    def __init__(self):
+        self.model = dict()
+        self.metrics = dict()
+
```
Comment (zhouqiang-cl) [inline oasis/models/model.py]: why need another init

File: oasis/models/rules.py [hunk 4]
```diff
@@ -2,15 +2,167 @@
...
+from oasis.models.model import Model, Config
 from oasis.datasource import PrometheusAPI, PrometheusQuery, Metrics
+from oasis.libs.log import logger
+from oasis.models.util import sub_id
+from oasis.libs.features import Features
+
+OPERATORS = ["==", "!=", "<=", ">=", "<", ">"]
 
 
 class Rules(Model):
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: I am a little confused by Rules. Is it just rules  or it is object contains rules. From the code it seems it runs job and map with rules.

File: oasis/models/rules.py [hunk 5]
```diff
@@ -2,15 +2,167 @@
...
+                           args=(key, val, self.cfg.metrics[key],
+                                 self.cfg.rules[key], ))
+                t.start()
+                self.threads[key] = t
+
+    def run_action(self, metric, val, config, rules):
+        logger.info("[job-id:{id}][metric:{metric}] start ot run"
+                    .format(id=sub_id(self.job.id), metric=metric))
+        self.compute(metric, val, config, rules)
+
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: split this function, It too long

File: oasis/models/rules.py [hunk 6]
```diff
@@ -2,15 +2,167 @@
...
+    def _set_default_config(self):
+        self.model.setdefault("data_range", "10m")
+        self.model.setdefault("predict_interval", "5m")
+
+        for metric in self.metrics:
+            self.metrics[metric].setdefault("features", ["mean", "std"])
+            self.metrics[metric].setdefault("rules",
+                                            ['features["std"] > 1000'])
+
+    def _parse_rules(self):
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: can we have some lib to parse rules

File: oasis/models/rules.py (hunk 4 above)
Comment (cwen0) [reply oasis/models/rules.py]: yep, this rules just define for metric, if the metric doesn't match the rule, we can judge the metric is not abnormal. this model just do it

File: oasis/models/model.py (hunk 3 above)
Comment (cwen0) [reply oasis/models/model.py]: abstract the same code and Config can be rewrited.

File: oasis/models/rules.py [hunk 7]
```diff
@@ -2,15 +2,167 @@
...
+def check_is_triggered(left_value, operator, right_value):
+    return {
+        "==": abs(left_value-right_value) < 0.0001,
+        "!=": abs(left_value-right_value) > 0.0001,
+        "<=": left_value <= right_value,
+        ">=": left_value >= right_value,
+        "=": abs(left_value-right_value) < 0.0001,
+        "<": left_value < right_value,
+        ">": left_value > right_value,
+    }.get(operator, False)
```
Comment (foreverbell) [inline oasis/models/rules.py]: log an error here if operator is unsupported

File: oasis/models/rules.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# support features: mean, std, var, sum, min, max
+# rules:
+# features["std"]
```
Comment (foreverbell) [inline oasis/models/rules.yml]: Be consistent. Here we use `feature["std"]`, why later switch to `feature[std]`?

File: oasis/models/iforest.yml [hunk 9]
```diff
@@ -1,7 +1,10 @@
+# support features: mean, std, var, sum, min, max
 model:
   train_count: 120
-  train_interval: 60
-  predict_interval: 300
+  train_data_range: 15m
```
Comment (foreverbell) [inline oasis/models/iforest.yml]: We should document these parameters at somewhere.

File: oasis/models/rules.py [hunk 10]
```diff
@@ -2,15 +2,167 @@
...
+            self.metrics[metric].setdefault("features", ["mean", "std"])
+            self.metrics[metric].setdefault("rules",
+                                            ['features["std"] > 1000'])
+
+    def _parse_rules(self):
+        for metric in self.metrics:
+            rules = self.metrics[metric].get("rules")
+            rules_list = []
+            for rule in rules:
+                index_left = rule.find("[")
```
Comment (foreverbell) [inline oasis/models/rules.py]: This logic is fragile. Can you define the grammar for rule?

Like `"feature[" feature "]" op value`?

File: oasis/models/rules.py [hunk 11]
```diff
@@ -2,15 +2,167 @@
...
+                if index_right == -1:
+                    logger.info("rule:{rule} is invalid, skip it"
+                                .format(rule=rule))
+
+                feature = rule[index_left+1:index_right]
+
+                operator = ""
+                operator_index = -1
+                for op in OPERATORS:
+                    operator_index = rule.find(op)
```
Comment (foreverbell) [inline oasis/models/rules.py]: Looks like `>= features[std] 0.1` is also a valid expression?

File: oasis/models/rules.py [hunk 12]
```diff
@@ -2,15 +2,167 @@
...
+
+                logger.info("[job-id:{id}][metric:{metric}] extraction features {value}, "
+                            "start to compare with rule"
+                            .format(id=sub_id(self.job.id),
+                                    metric=metric, value=features_value))
+
+                not_match_count = 0
+                for rule in rules:
+                    if rule["feature"] not in features_value:
+                        continue
```
Comment (foreverbell) [inline oasis/models/rules.py]: I think we should report an error eagerly about this during initialization.

File: oasis/models/rules.py [hunk 13]
```diff
@@ -2,15 +2,167 @@
...
+
+                    if check_is_triggered(features_value[rule["feature"]],
+                                          rule["operator"], rule["value"]):
+                        logger.info("[job-id:{id}][metric:{metric}] not match rule: {rule}"
+                                    .format(id=sub_id(self.job.id), metric=metric, rule=rule))
+                        self.callback("[job] {job}, metric: {metric} not match rule: {rule}"
+                                      .format(job=dict(self.job), metric=metric,
+                                              rule=rule), self.job.slack_channel)
+                        not_match_count = not_match_count + 1
```
Comment (foreverbell) [inline oasis/models/rules.py]: Looks like if `not_match_count > 0`, we can break early here. So, something like `match_flag` would be more appropriate?

File: oasis/models/iforest.py [hunk 14]
```diff
@@ -201,5 +181,6 @@ def _parse_api_config(self, config):
                 for key in config[section]:
                     self.metrics[key] = config[section][key]
 
-
+        for metric in self.metrics:
```
Comment (foreverbell) [inline oasis/models/iforest.py]: Move to `_set_default_config`.

File: oasis/models/model.py [hunk 15]
```diff
@@ -2,23 +2,66 @@
...
+        self._parse_config_file(config_file)
+
+        # load config from json data
+        if config is not None:
+            self._parse_api_config(config)
+
+        # set default config
+        self._set_default_config()
+
+    def _parse_config_file(self, config_file):
```
Comment (foreverbell) [inline oasis/models/model.py]: Looks like all `_parse_config_file`s are duplicate.

File: oasis/models/model.py [hunk 16]
```diff
@@ -2,23 +2,66 @@
...
+        with open(config_file, 'r') as config_file:
+            cfg = yaml.load(config_file)
+
+        self.model = cfg.get("model", dict())
+        self.metrics = cfg.get("metrics", dict())
+
+    def _parse_api_config(self, config):
```
Comment (foreverbell) [inline oasis/models/model.py]: Looks like all `_parse_api_config`s are mostly duplicate.

File: oasis/oasis.py [hunk 17]
```diff
@@ -92,8 +92,17 @@ class JobNewHandler(ModelHandler):
     @tornado.web.asynchronous
     @tornado.gen.coroutine
     def post(self):
-        data = json.loads(self.request.body)
-        print(data)
+        # data = dict()
```
Comment (foreverbell) [inline oasis/oasis.py]: remove this line

File: oasis/oasis.py [hunk 18]
```diff
@@ -92,8 +92,17 @@ class JobNewHandler(ModelHandler):
...
+        try:
+            data = json.loads(self.request.body)
+        except Exception as e:
+            logger.error("load json: {json_data}\nfailed: {err}"
+                         .format(json_data=self.request.body, err=e.__str__()))
```
Comment (foreverbell) [inline oasis/oasis.py]: Don't call `__str__` directly, use `str(e)`.

File: oasis/models/model.py [hunk 19]
```diff
@@ -2,23 +2,66 @@
...
+            for key in self.threads:
+                self.threads[key].join()
 
 
+class Config(object):
+    def __init__(self):
+        self.model = dict()
+        self.metrics = dict()
+
+    def _init(self, config_file, config=None):
```
[the remark this answers is not in the document]
Comment (foreverbell) [reply oasis/models/model.py]: I see... Better rename `config` to `config_api`, and comment the initialization priority.

.. and should we move `_set_default_config` to the first line for clarity? It has the least priority.

File: oasis/models/rules.py [hunk 20]
```diff
@@ -2,15 +2,167 @@
...
+                value_str = rule[operator_index+len(operator):]
+                value = float(value_str)
+                rules_list.append({"expr": rule, "feature": feature,
+                                   "operator": operator, "value": value})
 
+            self.rules[metric] = rules_list
 
 
+def check_is_triggered(left_value, operator, right_value):
+    return {
```
Comment (foreverbell) [inline oasis/models/rules.py]: suggestion: all keys are evaluated eagerly when the dict is initialized. This is costy if evaluation is expensive.

You can try something like `"==": lambda: abs(left_value-right_value) < 0.0001`, and `.get(operator, lambda: False)()`.

File: oasis/models/rules.py [hunk 21]
```diff
@@ -2,15 +2,167 @@
...
+        self._parse_rules()
+
+    def _set_default_config(self):
+        self.model.setdefault("data_range", "10m")
+        self.model.setdefault("predict_interval", "5m")
+
+        for metric in self.metrics:
+            self.metrics[metric].setdefault("features", ["mean", "std"])
+            self.metrics[metric].setdefault("rules",
+                                            ['features["std"] > 1000'])
```
Comment (foreverbell) [inline oasis/models/rules.py]: I think the quotes around `std` is unnecessary.

File: oasis/models/iforest.py [hunk 22]
```diff
@@ -132,55 +124,43 @@ def run(self):
...
                 val = Metrics[key]
-                config = {"features": ["mean", "std"]}
-                if key in self.cfg.metrics and "features" in self.cfg.metrics[key]:
-                    config = self.cfg.metrics[key]
+                if key not in self.cfg.metrics:
+                    continue
 
-                t = Thread(target=self.run_action, args=(key, val, config, ))
+                t = Thread(target=self.run_action,
+                           args=(key, val, self.cfg.metrics[key], ))
```
Comment (foreverbell) [inline oasis/models/iforest.py]: Remove the last comma, it is only needed if this tuple contains only one element

File: oasis/models/model.py (hunk 3 above)
Comment (foreverbell) [reply oasis/models/model.py]: Why not `def __init__(self, config_file, config=None):`?

**Comment by cwen0** [comment]
@zhouqiang-cl @foreverbell PTAL again

File: oasis/libs/rules.py [hunk 23]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+
+import re
+
+OPERATORS = ["==", "!=", "<=", ">=", "<", ">"]
+
+FEATURE_RULE = r'^features\[(\w+)\]\s*(==|!=|<=|>=|=|<|>)\s*(\d+|\.)'
```
Comment (foreverbell) [inline oasis/libs/rules.py]: '|'.join(OPERATORS)

File: oasis/libs/rules.py [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def parse_rule(rules):
+    rules_list = []
+    for rule in rules:
+        values = re.match(FEATURE_RULE, rule)
+        rules_list.append({"expr": rule, "feature": values.group(1),
+                           "operator": values.group(2),
+                           "value": float(values.group(3))})
```
Comment (foreverbell) [inline oasis/libs/rules.py]: caution: `float("1.2.3")` may raise an exception.

File: oasis/models/iforest.py [hunk 25]
```diff
@@ -7,32 +7,25 @@
 import pandas as pd
 import datetime
 import random
-import yaml
-import os
-from threading import Thread
 from oasis.datasource import PrometheusAPI, PrometheusQuery, Metrics
 from sklearn.ensemble import IsolationForest
-from threading import Event, Lock, Timer
+from threading import Timer, Thread
```
Comment (foreverbell) [inline oasis/models/iforest.py]: Can we get rid of `threading`? This may not be safe in tornado.

Try `tornado.ioloop.PeriodicCallback`?

File: oasis/models/iforest.py [hunk 26]
```diff
@@ -132,74 +124,47 @@ def run(self):
...
             if key in Metrics:
                 val = Metrics[key]
-                config = {"features": ["mean", "std"]}
-                if key in self.cfg.metrics and "features" in self.cfg.metrics[key]:
-                    config = self.cfg.metrics[key]
+                if key not in self.cfg.metrics:
+                    continue
 
-                t = Thread(target=self.run_action, args=(key, val, config, ))
+                t = Thread(target=self.run_action,
```
Comment (foreverbell) [inline oasis/models/iforest.py]: Can we also get rid of `Thread` by using facilities from tornado?

May be done in another PR, if you find this is difficult.

File: oasis/models/model.py [hunk 27]
```diff
@@ -2,23 +2,71 @@
...
 
+class Config(object):
+    def __init__(self, config_file, config_json=None):
+        self.model = dict()
+        self.metrics = dict()
+        self.config_file = config_file
+        self.config_json = config_json
+
+    def _init(self):
+        # set default config
```
Comment (foreverbell) [inline oasis/models/model.py]: Why not merging `_init_` into `__init__`?

File: oasis/oasis.py [hunk 28]
```diff
@@ -18,7 +18,7 @@
 HTTP_OK = 200
 
 runners = dict()
-lock = threading.Lock()
+lock = tornado.locks.Lock()
```
Comment (foreverbell) [inline oasis/oasis.py]: suggestion: better one PR does one thing. But as you have done it, we can keep it in this PR.

File: oasis/models/rules.py [hunk 29]
```diff
@@ -2,15 +2,153 @@
...
+        "=": lambda: abs(left_value-right_value) < EPS,
+        "<": lambda: left_value < right_value,
+        ">": lambda: left_value > right_value,
+    }
+
+    if operator not in operators:
+        logger.error("operator {} is invalid".format(operator))
+        return False
+
+    return operators[operator]
```
Comment (foreverbell) [inline oasis/models/rules.py]: `operators[operator]()`, otherwise we return a lambda here.

File: oasis/libs/rules.py (hunk 23 above)
Comment (foreverbell) [inline oasis/libs/rules.py]: Why not simply `std > 100`, removing `features`?

Comment (cwen0) [reply oasis/libs/rules.py]: the rules is not only `features rule` in the future，I just add `features` to diff them

File: oasis/models/iforest.py (hunk 26 above)
Comment (cwen0) [reply oasis/models/iforest.py]: ok, I will fix this problem in single pr

**Comment by cwen0** [comment]
@foreverbell @zhouqiang-cl PTAL again

File: oasis/models/rules.py [hunk 30]
```diff
@@ -2,15 +2,151 @@
...
+
+    def extraction_features(self, data_set, config):
+        values = []
+        for data in data_set.values():
+            values.append(float(data))
+
+        features_value = {}
+        for key in config["features"]:
+            if key in Features:
+                features_value[key] = Features[key](values)
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: call Features as func?, did we need to eval Features[key]

File: oasis/models/rules.py [hunk 31]
```diff
@@ -2,15 +2,151 @@
...
+
+        return features_value
+
+    def match_rules(self, metric, features_value, rules):
+        match_flag = True
+        for rule in rules:
+            if rule["feature"] not in features_value:
+                continue
+
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: why need check_is_triggered

File: oasis/models/rules.py [hunk 32]
```diff
@@ -2,15 +2,151 @@
...
+        for metric in self.metrics:
+            rules = self.metrics[metric].get("rules")
+            self.rules[metric] = parse_rule(rules)
+
+
+def check_is_triggered(left_value, operator, right_value):
+    if operator not in OPERATORS:
+        logger.error("operator {} is invalid".format(operator))
+        return False
+
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: a little wired. why need lambda

File: oasis/models/rules.py (hunk 30 above)
Comment (cwen0) [reply oasis/models/rules.py]: yep, `Features` is a dict of funcs

File: oasis/models/rules.py [hunk 33]
```diff
@@ -2,15 +2,151 @@
...
+                                 self.cfg.rules[key]))
+                t.start()
+                self.threads[key] = t
+
+    def run_action(self, metric, val, config, rules):
+        logger.info("[job-id:{id}][metric:{metric}] start ot run"
+                    .format(id=sub_id(self.job.id), metric=metric))
+        self.compute(metric, val, config, rules)
+
+    def compute(self, metric, query_expr, config, rules):
```
Comment (zhouqiang-cl) [inline oasis/models/rules.py]: add a comment for compute, I am confused about what data it want to confused, and return what data

File: oasis/models/rules.py (hunk 31 above)
Comment (cwen0) [reply oasis/models/rules.py]: to match rule

File: oasis/models/rules.py (hunk 32 above)
Comment (cwen0) [reply oasis/models/rules.py]: Make each expression as a whole.

**Comment by zhouqiang-cl** [comment]
I think I need some of your help to understand all of this code😄

**Comment by cwen0** [comment]
@zhouqiang-cl My fault, there are too few comments.
